Emit fewer `assume`s now that we have `range` metadata on parameters

scottmcm opened the pull request:
We still need the `assume` for the *target* type's range, but we no longer need it for the *source* type's range.

The first stab at this regressed a test, but thanks to good advice in https://github.com/llvm/llvm-project/issues/123278#issuecomment-2597440158 the second commit here changes how we emit these range assertions to the form that LLVM apparently likes better (and, conveniently, is easier to emit too) which got that test passing again 🎉 

Hopefully this means less crud for LLVM to churn through in `opt` builds...

scottmcm committed:
Emit fewer `assume`s now that we have `range` metadata on parameters

We still need the `assume` for the *target* type's range, but we no longer need it for the *source* type's range.

Admittedly there's one test not properly handled by LLVM today, but it's synthetic, so I'd still be fine doing this and just updating the test once LLVM fixes the bug.  All the other optimization tests still pass.

Hopefully this means less crud for LLVM to churn through in `opt` builds...

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 959fd5f1c7597fcaf2b76a838792467b2f28c85d with merge f0c078eba80206e6a26efad8143155ffd3c02ba8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"959fd5f1c7597fcaf2b76a838792467b2f28c85d","merge_sha":"f0c078eba80206e6a26efad8143155ffd3c02ba8"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12822565700/job/35755708110) [(plain)](https://github.com/rust-lang/rust/commit/959fd5f1c7597fcaf2b76a838792467b2f28c85d/checks/35755708110/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 27.6s done
#21 DONE 30.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
failures:

---- [codegen] tests/codegen/intrinsics/transmute-niched.rs#OPT stdout ----

error in revision `OPT`: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll" "/checkout/tests/codegen/intrinsics/transmute-niched.rs" "--check-prefix=CHECK" "--check-prefix" "OPT" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/codegen/intrinsics/transmute-niched.rs:34:14: error: OPT-SAME: expected string not found in input
/checkout/tests/codegen/intrinsics/transmute-niched.rs:34:14: error: OPT-SAME: expected string not found in input
// OPT-SAME: range(i8 10, 13){{.+}}%x
             ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll:17:36: note: scanning from here
define noundef i8 @check_from_enum(i8 noundef %x) unnamed_addr #0 {
                                   ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll:19:2: note: possible intended match here
 ret i8 %x
/checkout/tests/codegen/intrinsics/transmute-niched.rs:61:14: error: OPT-SAME: expected string not found in input
/checkout/tests/codegen/intrinsics/transmute-niched.rs:61:14: error: OPT-SAME: expected string not found in input
// OPT-SAME: range(i8 -1, 2){{.+}}%x
             ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll:33:40: note: scanning from here
define noundef i8 @check_from_ordering(i8 noundef %x) unnamed_addr #0 {
                                       ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll:35:2: note: possible intended match here
 ret i8 %x
/checkout/tests/codegen/intrinsics/transmute-niched.rs:97:14: error: OPT-SAME: expected string not found in input
/checkout/tests/codegen/intrinsics/transmute-niched.rs:97:14: error: OPT-SAME: expected string not found in input
// OPT-SAME: range(i32 0, 1114112){{.+}}%x
             ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll:39:42: note: scanning from here
define noundef i32 @check_enum_from_char(i32 noundef %x) unnamed_addr #0 {
                                         ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll:41:10: note: possible intended match here
 %0 = icmp uge i32 %x, -100
/checkout/tests/codegen/intrinsics/transmute-niched.rs:114:14: error: OPT-SAME: expected string not found in input
/checkout/tests/codegen/intrinsics/transmute-niched.rs:114:14: error: OPT-SAME: expected string not found in input
// OPT-SAME: range(i32 -100, 101){{.+}}%x
             ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll:49:40: note: scanning from here
define noundef i32 @check_enum_to_char(i32 noundef %x) unnamed_addr #0 {
                                       ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll:51:10: note: possible intended match here
 %0 = icmp ule i32 %x, 1114111
/checkout/tests/codegen/intrinsics/transmute-niched.rs:129:14: error: OPT-SAME: expected string not found in input
/checkout/tests/codegen/intrinsics/transmute-niched.rs:129:14: error: OPT-SAME: expected string not found in input
// OPT-SAME: range(i32 0, 1114112){{.+}}%x.0
             ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll:57:38: note: scanning from here
define { i32, i32 } @check_swap_pair(i32 noundef %x.0, i32 noundef %x.1) unnamed_addr #0 {
                                     ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll:61:10: note: possible intended match here
 %1 = icmp ule i32 %x.1, 1114111
/checkout/tests/codegen/intrinsics/transmute-niched.rs:149:14: error: OPT-SAME: expected string not found in input
/checkout/tests/codegen/intrinsics/transmute-niched.rs:149:14: error: OPT-SAME: expected string not found in input
// OPT-SAME: range(i8 -1, 2){{.+}}%x
             ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll:69:53: note: scanning from here
define noundef zeroext i1 @check_bool_from_ordering(i8 noundef %x) unnamed_addr #0 {
                                                    ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll:71:10: note: possible intended match here
 %0 = icmp ule i8 %x, 1

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/intrinsics/transmute-niched.OPT/transmute-niched.ll
Check file: /checkout/tests/codegen/intrinsics/transmute-niched.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'transmute_niched.773cff602b7825b4-cgu.0' 
            2: source_filename = "transmute_niched.773cff602b7825b4-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
            5:  
            5:  
            6: ; Function Attrs: nonlazybind uwtable 
            7: define noundef i8 @check_to_enum(i8 noundef %x) unnamed_addr #0 { 
            8: start: 
            9:  %0 = icmp uge i8 %x, 10 
           10:  call void @llvm.assume(i1 %0) 
           11:  %1 = icmp ule i8 %x, 12 
           12:  call void @llvm.assume(i1 %1) 
           13:  ret i8 %x 
           14: } 
           15:  
           16: ; Function Attrs: nonlazybind uwtable 
           17: define noundef i8 @check_from_enum(i8 noundef %x) unnamed_addr #0 { 
same:34'0                                         X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           18: start: 
same:34'0      ~~~~~~~
           19:  ret i8 %x 
same:34'0      ~~~~~~~~~~~
same:34'1       ?          possible intended match
           20: } 
same:34'0      ~~
           21:  
same:34'0      ~
           22: ; Function Attrs: nonlazybind uwtable 
same:34'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           23: define noundef i8 @check_to_ordering(i8 noundef %x) unnamed_addr #0 { 
same:34'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           24: start: 
           25:  %0 = icmp uge i8 %x, -1 
           26:  %1 = icmp ule i8 %x, 1 
           27:  %2 = or i1 %0, %1 
           28:  call void @llvm.assume(i1 %2) 
           29:  ret i8 %x 
           30: } 
           31:  
           32: ; Function Attrs: nonlazybind uwtable 
           33: define noundef i8 @check_from_ordering(i8 noundef %x) unnamed_addr #0 { 
same:61'0                                             X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           34: start: 
same:61'0      ~~~~~~~
           35:  ret i8 %x 
same:61'0      ~~~~~~~~~~~
same:61'1       ?          possible intended match
           36: } 
same:61'0      ~~
           37:  
same:61'0      ~
           38: ; Function Attrs: nonlazybind uwtable 
same:61'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           39: define noundef i32 @check_enum_from_char(i32 noundef %x) unnamed_addr #0 { 
same:61'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
same:97'0                                               X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           40: start: 
same:97'0      ~~~~~~~
           41:  %0 = icmp uge i32 %x, -100 
same:97'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
same:97'1               ?                   possible intended match
           42:  %1 = icmp ule i32 %x, 100 
same:97'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           43:  %2 = or i1 %0, %1 
same:97'0      ~~~~~~~~~~~~~~~~~~~
           44:  call void @llvm.assume(i1 %2) 
same:97'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           45:  ret i32 %x 
same:97'0      ~~~~~~~~~~~~
           46: } 
same:97'0      ~~
           47:  
same:97'0      ~
           48: ; Function Attrs: nonlazybind uwtable 
same:97'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           49: define noundef i32 @check_enum_to_char(i32 noundef %x) unnamed_addr #0 { 
same:97'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
same:114'0                                            X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           50: start: 
same:114'0     ~~~~~~~
           51:  %0 = icmp ule i32 %x, 1114111 
same:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
same:114'1              ?                      possible intended match
           52:  call void @llvm.assume(i1 %0) 
same:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           53:  ret i32 %x 
same:114'0     ~~~~~~~~~~~~
           54: } 
same:114'0     ~~
           55:  
same:114'0     ~
           56: ; Function Attrs: nonlazybind uwtable 
same:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           57: define { i32, i32 } @check_swap_pair(i32 noundef %x.0, i32 noundef %x.1) unnamed_addr #0 { 
same:114'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
same:129'0                                          X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           58: start: 
same:129'0     ~~~~~~~
           59:  %0 = icmp uge i32 %x.0, 1 
same:129'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           60:  call void @llvm.assume(i1 %0) 
same:129'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           61:  %1 = icmp ule i32 %x.1, 1114111 
same:129'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
same:129'1              ?                        possible intended match
           62:  call void @llvm.assume(i1 %1) 
same:129'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           63:  %2 = insertvalue { i32, i32 } poison, i32 %x.0, 0 
same:129'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           64:  %3 = insertvalue { i32, i32 } %2, i32 %x.1, 1 
same:129'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           65:  ret { i32, i32 } %3 
same:129'0     ~~~~~~~~~~~~~~~~~~~~~
           66: } 
same:129'0     ~~
           67:  
same:129'0     ~
           68: ; Function Attrs: nonlazybind uwtable 
same:129'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           69: define noundef zeroext i1 @check_bool_from_ordering(i8 noundef %x) unnamed_addr #0 { 
same:129'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
same:149'0                                                         X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           70: start: 
same:149'0     ~~~~~~~
           71:  %0 = icmp ule i8 %x, 1 
same:149'0     ~~~~~~~~~~~~~~~~~~~~~~~~
same:149'1              ?               possible intended match
           72:  call void @llvm.assume(i1 %0) 
same:149'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           73:  %_0 = trunc i8 %x to i1 
same:149'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
           74:  ret i1 %_0 
same:149'0     ~~~~~~~~~~~~
           75: } 
same:149'0     ~~
           76:  
same:149'0     ~
           77: ; Function Attrs: nonlazybind uwtable 
same:149'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           78: define noundef i8 @check_bool_to_ordering(i1 noundef zeroext %x) unnamed_addr #0 { 
same:149'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           79: start: 
           80:  %_0 = zext i1 %x to i8 
           81:  %0 = icmp uge i8 %_0, -1 
           82:  %1 = icmp ule i8 %_0, 1 
           83:  %2 = or i1 %0, %1 
           84:  call void @llvm.assume(i1 %2) 
           85:  ret i8 %_0 
           86: } 
           87:  
           88: ; Function Attrs: nocallback nofree nosync nounwind willreturn memory(inaccessiblemem: write) 
           89: declare void @llvm.assume(i1 noundef) #1 
           90:  
           91: attributes #0 = { nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" "target-features"="+x87,+sse2" } 
           92: attributes #1 = { nocallback nofree nosync nounwind willreturn memory(inaccessiblemem: write) } 
           93:  
           94: !llvm.module.flags = !{!0, !1} 
           95: !llvm.ident = !{!2} 
           96:  
           97: !0 = !{i32 8, !"PIC Level", i32 2} 
           98: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
           99: !2 = !{!"rustc version 1.86.0-nightly (3ea7c244d 2025-01-17)"} 
------------------------------------------


---- [codegen] tests/codegen/issues/issue-119422.rs stdout ----
---- [codegen] tests/codegen/issues/issue-119422.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-119422/issue-119422.ll" "/checkout/tests/codegen/issues/issue-119422.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/codegen/issues/issue-119422.rs:22:12: error: CHECK: expected string not found in input
/checkout/tests/codegen/issues/issue-119422.rs:22:12: error: CHECK: expected string not found in input
 // CHECK: ret i1 false
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-119422/issue-119422.ll:13:51: note: scanning from here
define noundef zeroext i1 @equals_zero_is_false_u8(i8 noundef %x) unnamed_addr #0 {
                                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-119422/issue-119422.ll:16:2: note: possible intended match here
 ret i1 %_0
/checkout/tests/codegen/issues/issue-119422.rs:31:12: error: CHECK: expected string not found in input
/checkout/tests/codegen/issues/issue-119422.rs:31:12: error: CHECK: expected string not found in input
 // CHECK: ret i1 true
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-119422/issue-119422.ll:20:54: note: scanning from here
define noundef zeroext i1 @not_equals_zero_is_true_u8(i8 noundef %x) unnamed_addr #0 {
                                                     ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-119422/issue-119422.ll:23:2: note: possible intended match here
 ret i1 %_0
/checkout/tests/codegen/issues/issue-119422.rs:40:12: error: CHECK: expected string not found in input
/checkout/tests/codegen/issues/issue-119422.rs:40:12: error: CHECK: expected string not found in input
 // CHECK: ret i1 false
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-119422/issue-119422.ll:27:51: note: scanning from here
define noundef zeroext i1 @equals_zero_is_false_i8(i8 noundef %x) unnamed_addr #0 {
                                                  ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-119422/issue-119422.ll:30:2: note: possible intended match here
 ret i1 %_0
/checkout/tests/codegen/issues/issue-119422.rs:49:12: error: CHECK: expected string not found in input
/checkout/tests/codegen/issues/issue-119422.rs:49:12: error: CHECK: expected string not found in input
 // CHECK: ret i1 true
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-119422/issue-119422.ll:34:54: note: scanning from here
define noundef zeroext i1 @not_equals_zero_is_true_i8(i8 noundef %x) unnamed_addr #0 {
                                                     ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-119422/issue-119422.ll:37:2: note: possible intended match here
 ret i1 %_0
/checkout/tests/codegen/issues/issue-119422.rs:78:12: error: CHECK: expected string not found in input
/checkout/tests/codegen/issues/issue-119422.rs:78:12: error: CHECK: expected string not found in input
 // CHECK: ret i1 false
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-119422/issue-119422.ll:59:56: note: scanning from here
define noundef zeroext i1 @u64_from_nonzero_is_not_zero(i64 noundef %x) unnamed_addr #0 {
                                                       ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-119422/issue-119422.ll:62:2: note: possible intended match here
 ret i1 %_0

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-119422/issue-119422.ll
Check file: /checkout/tests/codegen/issues/issue-119422.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'issue_119422.1ea4dfb3f5cd3f94-cgu.0' 
            2: source_filename = "issue_119422.1ea4dfb3f5cd3f94-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
            5:  
            5:  
            6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
            7: define noundef zeroext i1 @check_non_null(ptr nocapture noundef nonnull readnone %x) unnamed_addr #0 { 
            9:  ret i1 false 
           10: } 
           11:  
           11:  
           12: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
           13: define noundef zeroext i1 @equals_zero_is_false_u8(i8 noundef %x) unnamed_addr #0 { 
check:22'0                                                       X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
check:22'0     ~~~~~~~
check:22'0     ~~~~~~~
           15:  %_0 = icmp eq i8 %x, 0 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~
           16:  ret i1 %_0 
check:22'0     ~~~~~~~~~~~~
check:22'1      ?           possible intended match
           17: } 
check:22'0     ~~
check:22'0     ~
check:22'0     ~
           19: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           20: define noundef zeroext i1 @not_equals_zero_is_true_u8(i8 noundef %x) unnamed_addr #0 { 
check:22'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:31'0                                                          X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
check:31'0     ~~~~~~~
check:31'0     ~~~~~~~
           22:  %_0 = icmp ne i8 %x, 0 
check:31'0     ~~~~~~~~~~~~~~~~~~~~~~~~
           23:  ret i1 %_0 
check:31'0     ~~~~~~~~~~~~
check:31'1      ?           possible intended match
           24: } 
check:31'0     ~~
check:31'0     ~
check:31'0     ~
           26: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
check:31'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           27: define noundef zeroext i1 @equals_zero_is_false_i8(i8 noundef %x) unnamed_addr #0 { 
check:31'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:40'0                                                       X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
check:40'0     ~~~~~~~
check:40'0     ~~~~~~~
           29:  %_0 = icmp eq i8 %x, 0 
check:40'0     ~~~~~~~~~~~~~~~~~~~~~~~~
           30:  ret i1 %_0 
check:40'0     ~~~~~~~~~~~~
check:40'1      ?           possible intended match
           31: } 
check:40'0     ~~
check:40'0     ~
check:40'0     ~
           33: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
check:40'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           34: define noundef zeroext i1 @not_equals_zero_is_true_i8(i8 noundef %x) unnamed_addr #0 { 
check:40'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:49'0                                                          X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
check:49'0     ~~~~~~~
check:49'0     ~~~~~~~
           36:  %_0 = icmp ne i8 %x, 0 
check:49'0     ~~~~~~~~~~~~~~~~~~~~~~~~
           37:  ret i1 %_0 
check:49'0     ~~~~~~~~~~~~
check:49'1      ?           possible intended match
           38: } 
check:49'0     ~~
check:49'0     ~
check:49'0     ~
           40: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(inaccessiblemem: write) uwtable 
check:49'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           41: define noundef i64 @usize_try_from_u32(i32 noundef %x) unnamed_addr #1 personality ptr @rust_eh_personality { 
check:49'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           42: "_ZN4core6result19Result$LT$T$C$E$GT$6unwrap17h003751d47e9effc6E.exit": 
           43:  %_5 = zext i32 %x to i64 
           44:  %0 = icmp ne i32 %x, 0 
           45:  tail call void @llvm.assume(i1 %0) 
           46:  ret i64 %_5 
           47: } 
           48:  
           49: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(inaccessiblemem: write) uwtable 
           50: define noundef i64 @isize_try_from_i32(i32 noundef %x) unnamed_addr #1 personality ptr @rust_eh_personality { 
           51: "_ZN4core6result19Result$LT$T$C$E$GT$6unwrap17hf60c3c8760a9fe03E.exit": 
           52:  %_5 = sext i32 %x to i64 
           53:  %0 = icmp ne i32 %x, 0 
           54:  tail call void @llvm.assume(i1 %0) 
           55:  ret i64 %_5 
           56: } 
           57:  
           58: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
           59: define noundef zeroext i1 @u64_from_nonzero_is_not_zero(i64 noundef %x) unnamed_addr #0 { 
check:78'0                                                            X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
check:78'0     ~~~~~~~
check:78'0     ~~~~~~~
           61:  %_0 = icmp eq i64 %x, 0 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
           62:  ret i1 %_0 
check:78'0     ~~~~~~~~~~~~
check:78'1      ?           possible intended match
           63: } 
check:78'0     ~~
check:78'0     ~
check:78'0     ~
           65: ; Function Attrs: nounwind nonlazybind uwtable 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           66: declare noundef i32 @rust_eh_personality(i32 noundef, i32 noundef, i64 noundef, ptr noundef, ptr noundef) unnamed_addr #2 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:78'0     ~
check:78'0     ~
           68: ; Function Attrs: nocallback nofree nosync nounwind willreturn memory(inaccessiblemem: write) 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           69: declare void @llvm.assume(i1 noundef) #3 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:78'0     ~
check:78'0     ~
           71: attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" "target-features"="+x87,+sse2" } 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           72: attributes #1 = { mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(inaccessiblemem: write) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" "target-features"="+x87,+sse2" } 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           73: attributes #2 = { nounwind nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" "target-features"="+x87,+sse2" } 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           74: attributes #3 = { nocallback nofree nosync nounwind willreturn memory(inaccessiblemem: write) } 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:78'0     ~
check:78'0     ~
           76: !llvm.module.flags = !{!0, !1} 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           77: !llvm.ident = !{!2} 
check:78'0     ~~~~~~~~~~~~~~~~~~~~
check:78'0     ~
check:78'0     ~
           79: !0 = !{i32 8, !"PIC Level", i32 2} 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           80: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           81: !2 = !{!"rustc version 1.86.0-nightly (3ea7c244d 2025-01-17)"} 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------


---- [codegen] tests/codegen/transmute-optimized.rs stdout ----
---- [codegen] tests/codegen/transmute-optimized.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/transmute-optimized/transmute-optimized.ll" "/checkout/tests/codegen/transmute-optimized.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/codegen/transmute-optimized.rs:15:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 0
           ^
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/transmute-optimized/transmute-optimized.ll:11:37: note: scanning from here
define noundef i32 @issue_109958(i32 noundef returned %x) unnamed_addr #0 {
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/transmute-optimized/transmute-optimized.ll:13:2: note: possible intended match here
 ret i32 %x
 ^
/checkout/tests/codegen/transmute-optimized.rs:38:12: error: CHECK: expected string not found in input
/checkout/tests/codegen/transmute-optimized.rs:38:12: error: CHECK: expected string not found in input
 // CHECK: ret i1 false
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/transmute-optimized/transmute-optimized.ll:29:45: note: scanning from here
define noundef zeroext i1 @non_zero_is_null(i64 noundef %x) unnamed_addr #0 {
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/transmute-optimized/transmute-optimized.ll:32:2: note: possible intended match here
 ret i1 %_0
 ^
/checkout/tests/codegen/transmute-optimized.rs:80:16: error: CHECK-NOT: excluded string found in input
/checkout/tests/codegen/transmute-optimized.rs:80:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: call core::panicking::panic
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/transmute-optimized/transmute-optimized.ll:81:3: note: found here
; call core::panicking::panic_const::panic_const_div_by_zero
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/tests/codegen/transmute-optimized.rs:93:16: error: CHECK-SAME: expected string not found in input
// CHECK-SAME: range(i8 -1, 2){{.+}}%x
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/transmute-optimized/transmute-optimized.ll:87:53: note: scanning from here
define noundef i8 @ordering_transmute_onetwothree(i8 noundef returned %x) unnamed_addr #2 {
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/transmute-optimized/transmute-optimized.ll:89:10: note: possible intended match here
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/transmute-optimized/transmute-optimized.ll:89:10: note: possible intended match here
 %0 = icmp ne i8 %x, 0
/checkout/tests/codegen/transmute-optimized.rs:111:12: error: CHECK: expected string not found in input
 // CHECK: ret i8 1
           ^
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/transmute-optimized/transmute-optimized.ll:97:53: note: scanning from here
define noundef i8 @onetwothree_transmute_ordering(i8 noundef returned %x) unnamed_addr #2 {
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/transmute-optimized/transmute-optimized.ll:102:2: note: possible intended match here
 ret i8 %x
 ^
/checkout/tests/codegen/transmute-optimized.rs:118:12: error: CHECK: expected string not found in input
/checkout/tests/codegen/transmute-optimized.rs:118:12: error: CHECK: expected string not found in input
 // CHECK: ret i1 false
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/transmute-optimized/transmute-optimized.ll:106:48: note: scanning from here
define noundef zeroext i1 @char_is_negative(i32 noundef %c) unnamed_addr #0 {
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/transmute-optimized/transmute-optimized.ll:109:2: note: possible intended match here
 ret i1 %_0
 ^

```

</details>

scottmcm committed:
Emit range assumes as a single `icmp`

Thank you dtcxzyw in LLVM 123278 for pointing out that we were doing this in a suboptimal way.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12822582182/job/35758454370)
Build commit: f0c078eba80206e6a26efad8143155ffd3c02ba8 (`f0c078eba80206e6a26efad8143155ffd3c02ba8`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12822582182/job/35758454370"},"merge_sha":"f0c078eba80206e6a26efad8143155ffd3c02ba8"} -->

rust-timer commented:
Queued f0c078eba80206e6a26efad8143155ffd3c02ba8 with parent 76a030a6c2dab11760cb08d746fa515e4bce5d39, future [comparison URL](https://perf.rust-lang.org/compare.html?start=76a030a6c2dab11760cb08d746fa515e4bce5d39&end=f0c078eba80206e6a26efad8143155ffd3c02ba8).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (f0c078eba80206e6a26efad8143155ffd3c02ba8): [comparison URL](https://perf.rust-lang.org/compare.html?start=76a030a6c2dab11760cb08d746fa515e4bce5d39&end=f0c078eba80206e6a26efad8143155ffd3c02ba8&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=76a030a6c2dab11760cb08d746fa515e4bce5d39&end=f0c078eba80206e6a26efad8143155ffd3c02ba8&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.3% | [0.3%, 0.3%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=76a030a6c2dab11760cb08d746fa515e4bce5d39&end=f0c078eba80206e6a26efad8143155ffd3c02ba8&stat=max-rss)
<details>
<summary>Results (primary -0.6%, secondary 1.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.5%  | [1.5%, 1.5%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.5%  | [2.2%, 2.7%]   | 2     |
| Improvements ✅ <br /> (primary)   | -1.6% | [-2.3%, -1.0%] | 2     |
| Improvements ✅ <br /> (secondary) | -0.8% | [-0.8%, -0.8%] | 1     |
| All ❌✅ (primary)                 | -0.6% | [-2.3%, 1.5%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=76a030a6c2dab11760cb08d746fa515e4bce5d39&end=f0c078eba80206e6a26efad8143155ffd3c02ba8&stat=cycles:u)
<details>
<summary>Results (primary -0.8%, secondary 3.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.3%  | [2.0%, 4.9%]   | 4     |
| Improvements ✅ <br /> (primary)   | -0.8% | [-0.8%, -0.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.8% | [-0.8%, -0.8%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=76a030a6c2dab11760cb08d746fa515e4bce5d39&end=f0c078eba80206e6a26efad8143155ffd3c02ba8&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.0%, secondary 0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0%  | [0.0%, 0.0%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.0%  | [0.0%, 0.0%]   | 11    |
| Improvements ✅ <br /> (primary)   | -0.0% | [-0.0%, -0.0%] | 3     |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 1     |
| All ❌✅ (primary)                 | -0.0% | [-0.0%, 0.0%]  | 4     |
</details>

**Bootstrap**: 765.257s -> 763.281s (-0.26%)
**Artifact size**: 326.05 MiB -> 326.07 MiB (0.01%)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12825425491/job/35763370723) [(plain)](https://github.com/rust-lang/rust/commit/8ab474a9b9759be27b6bb23eba74749fb2feda5c/checks/35763370723/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 27.0s done
#21 DONE 39.2s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
failures:

---- [codegen] tests/codegen/cast-optimized.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/cast-optimized/cast-optimized.ll" "/checkout/tests/codegen/cast-optimized.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/codegen/cast-optimized.rs:29:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: call core::panicking::panic
               ^
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/cast-optimized/cast-optimized.ll:55:3: note: found here
; call core::panicking::panic_bounds_check
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/cast-optimized/cast-optimized.ll
Check file: /checkout/tests/codegen/cast-optimized.rs

-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
        1: ; ModuleID = 'cast_optimized.e443adfe031a928a-cgu.0' 
        2: source_filename = "cast_optimized.e443adfe031a928a-cgu.0" 
        3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
        5:  
        5:  
        6: @alloc_6355dbdd1624ffcc2c33b964d3397827 = private unnamed_addr constant <{ [41 x i8] }> <{ [41 x i8] c"/checkout/tests/codegen/cast-optimized.rs" }>, align 1 
        7: @alloc_605675d4041f71da95d6212c2817bd88 = private unnamed_addr constant <{ [8 x i8], ptr, [8 x i8] }> <{ [8 x i8] c"\0F\00\00\00\05\00\00\00", ptr @alloc_6355dbdd1624ffcc2c33b964d3397827, [8 x i8] c")\00\00\00\00\00\00\00" }>, align 8 
        8: @alloc_ded7e7b8673d6b9aac84802a4c41cc23 = private unnamed_addr constant <{ [8 x i8], ptr, [8 x i8] }> <{ [8 x i8] c"\1E\00\00\00\05\00\00\00", ptr @alloc_6355dbdd1624ffcc2c33b964d3397827, [8 x i8] c")\00\00\00\00\00\00\00" }>, align 8 
        9:  
       10: ; Function Attrs: nonlazybind uwtable 
       11: define void @u32_index(ptr dead_on_unwind noalias nocapture noundef writable writeonly sret([22 x i8]) align 1 dereferenceable(22) %_0, i32 noundef %c) unnamed_addr #0 { 
       12: start: 
       13:  %array = alloca [22 x i8], align 1 
       14:  call void @llvm.lifetime.start.p0(i64 22, ptr nonnull %array) 
       15:  call void @llvm.memset.p0.i64(ptr noundef nonnull align 1 dereferenceable(22) %array, i8 0, i64 22, i1 false) 
       16:  %0 = tail call i32 @llvm.ctlz.i32(i32 %c, i1 false), !range !3 
       17:  %index = sub nuw nsw i32 32, %0 
       18:  %_5 = zext nneg i32 %index to i64 
       19:  %_6 = icmp ult i32 %c, 2097152 
       20:  br i1 %_6, label %bb1, label %panic 
       21:  
       22: bb1: ; preds = %start 
       23:  %1 = getelementptr inbounds i8, ptr %array, i64 %_5 
       24:  store i8 1, ptr %1, align 1 
       25:  call void @llvm.memcpy.p0.p0.i64(ptr noundef nonnull align 1 dereferenceable(22) %_0, ptr noundef nonnull align 1 dereferenceable(22) %array, i64 22, i1 false) 
       26:  call void @llvm.lifetime.end.p0(i64 22, ptr nonnull %array) 
       27:  ret void 
       29: panic: ; preds = %start 
       30: ; call core::panicking::panic_bounds_check 
       30: ; call core::panicking::panic_bounds_check 
       31:  tail call void @_RNvNtCsjueXZ9qMhAP_4core9panicking18panic_bounds_check(i64 noundef %_5, i64 noundef 22, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_605675d4041f71da95d6212c2817bd88) #6 
       32:  unreachable 
       33: } 
       34:  
       35: ; Function Attrs: nonlazybind uwtable 
       36: define void @char_as_u32_index(ptr dead_on_unwind noalias nocapture noundef writable writeonly sret([22 x i8]) align 1 dereferenceable(22) %_0, i32 noundef %c) unnamed_addr #0 { 
       37: start: 
       38:  %array = alloca [22 x i8], align 1 
       39:  call void @llvm.lifetime.start.p0(i64 22, ptr nonnull %array) 
       40:  call void @llvm.memset.p0.i64(ptr noundef nonnull align 1 dereferenceable(22) %array, i8 0, i64 22, i1 false) 
       41:  %0 = tail call i32 @llvm.ctlz.i32(i32 %c, i1 false), !range !3 
       42:  %index = sub nuw nsw i32 32, %0 
       43:  %_6 = zext nneg i32 %index to i64 
       44:  %_7 = icmp ult i32 %c, 2097152 
       45:  br i1 %_7, label %bb1, label %panic 
       46:  
       47: bb1: ; preds = %start 
       48:  %1 = getelementptr inbounds i8, ptr %array, i64 %_6 
       49:  store i8 1, ptr %1, align 1 
       50:  call void @llvm.memcpy.p0.p0.i64(ptr noundef nonnull align 1 dereferenceable(22) %_0, ptr noundef nonnull align 1 dereferenceable(22) %array, i64 22, i1 false) 
       51:  call void @llvm.lifetime.end.p0(i64 22, ptr nonnull %array) 
       52:  ret void 
       54: panic: ; preds = %start 
       55: ; call core::panicking::panic_bounds_check 
       55: ; call core::panicking::panic_bounds_check 
not:29       !~~~~~~~~~~~~~~~~~~~~~~~~~~               error: no match expected
       56:  tail call void @_RNvNtCsjueXZ9qMhAP_4core9panicking18panic_bounds_check(i64 noundef %_6, i64 noundef 22, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_ded7e7b8673d6b9aac84802a4c41cc23) #6 
       57:  unreachable 
       58: } 
       59:  
       60: ; Function Attrs: mustprogress nocallback nofree nounwind willreturn memory(argmem: write) 
       61: declare void @llvm.memset.p0.i64(ptr nocapture writeonly, i8, i64, i1 immarg) #1 
       62:  
       63: ; Function Attrs: mustprogress nocallback nofree nosync nounwind speculatable willreturn memory(none) 
       64: declare i32 @llvm.ctlz.i32(i32, i1 immarg) #2 
       66: ; core::panicking::panic_bounds_check 
       66: ; core::panicking::panic_bounds_check 
       67: ; Function Attrs: cold minsize noinline noreturn nonlazybind optsize uwtable 
       68: declare void @_RNvNtCsjueXZ9qMhAP_4core9panicking18panic_bounds_check(i64 noundef, i64 noundef, ptr noalias noundef readonly align 8 dereferenceable(24)) unnamed_addr #3 
       69:  
       70: ; Function Attrs: mustprogress nocallback nofree nounwind willreturn memory(argmem: readwrite) 
       71: declare void @llvm.memcpy.p0.p0.i64(ptr noalias nocapture writeonly, ptr noalias nocapture readonly, i64, i1 immarg) #4 
       72:  
       73: ; Function Attrs: mustprogress nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
       74: declare void @llvm.lifetime.start.p0(i64 immarg, ptr nocapture) #5 
       75:  
       76: ; Function Attrs: mustprogress nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
       77: declare void @llvm.lifetime.end.p0(i64 immarg, ptr nocapture) #5 
       78:  
       79: attributes #0 = { nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" "target-features"="+x87,+sse2" } 
       80: attributes #1 = { mustprogress nocallback nofree nounwind willreturn memory(argmem: write) } 
       81: attributes #2 = { mustprogress nocallback nofree nosync nounwind speculatable willreturn memory(none) } 
       82: attributes #3 = { cold minsize noinline noreturn nonlazybind optsize uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" "target-features"="+x87,+sse2" } 
       83: attributes #4 = { mustprogress nocallback nofree nounwind willreturn memory(argmem: readwrite) } 
       84: attributes #5 = { mustprogress nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) } 
       85: attributes #6 = { noreturn } 
       86:  
       87: !llvm.module.flags = !{!0, !1} 
       88: !llvm.ident = !{!2} 
       89:  
       90: !0 = !{i32 8, !"PIC Level", i32 2} 
       91: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
       92: !2 = !{!"rustc version 1.86.0-nightly (fde38769d 2025-01-17)"} 
       93: !3 = !{i32 0, i32 33} 
------------------------------------------



```

</details>

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 8ab474a9b9759be27b6bb23eba74749fb2feda5c with merge 28bc5054392a5de43833561ce0416a1f8cfdf227...
<!-- homu: {"type":"TryBuildStarted","head_sha":"8ab474a9b9759be27b6bb23eba74749fb2feda5c","merge_sha":"28bc5054392a5de43833561ce0416a1f8cfdf227"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12835026205/job/35798138378)
Build commit: 28bc5054392a5de43833561ce0416a1f8cfdf227 (`28bc5054392a5de43833561ce0416a1f8cfdf227`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12835026205/job/35798138378"},"merge_sha":"28bc5054392a5de43833561ce0416a1f8cfdf227"} -->

rust-timer commented:
Queued 28bc5054392a5de43833561ce0416a1f8cfdf227 with parent bcd0683e5dce1945b5d940714742e7502883bb5c, future [comparison URL](https://perf.rust-lang.org/compare.html?start=bcd0683e5dce1945b5d940714742e7502883bb5c&end=28bc5054392a5de43833561ce0416a1f8cfdf227).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (28bc5054392a5de43833561ce0416a1f8cfdf227): [comparison URL](https://perf.rust-lang.org/compare.html?start=bcd0683e5dce1945b5d940714742e7502883bb5c&end=28bc5054392a5de43833561ce0416a1f8cfdf227&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=bcd0683e5dce1945b5d940714742e7502883bb5c&end=28bc5054392a5de43833561ce0416a1f8cfdf227&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.8% | [0.4%, 1.7%] | 6     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=bcd0683e5dce1945b5d940714742e7502883bb5c&end=28bc5054392a5de43833561ce0416a1f8cfdf227&stat=max-rss)
<details>
<summary>Results (primary -4.5%, secondary -1.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -4.5% | [-4.5%, -4.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.0% | [-1.0%, -1.0%] | 1     |
| All ❌✅ (primary)                 | -4.5% | [-4.5%, -4.5%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=bcd0683e5dce1945b5d940714742e7502883bb5c&end=28bc5054392a5de43833561ce0416a1f8cfdf227&stat=cycles:u)
<details>
<summary>Results (secondary -2.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.3% | [-2.3%, -2.2%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=bcd0683e5dce1945b5d940714742e7502883bb5c&end=28bc5054392a5de43833561ce0416a1f8cfdf227&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.0%, secondary 0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0%  | [0.0%, 0.0%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 0.0%  | [0.0%, 0.0%]   | 15    |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.1%, -0.1%] | 3     |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 1     |
| All ❌✅ (primary)                 | -0.0% | [-0.1%, 0.0%]  | 4     |
</details>

**Bootstrap**: 763.784s -> 764.347s (0.07%)
**Artifact size**: 326.10 MiB -> 326.07 MiB (-0.01%)
